p2panda-rs/src/secret_group/tests.rs

```diff
@@ -0,0 +1,234 @@
+        .add_members(&billie_provider, &[ada_key_package])
+        .unwrap();
+    assert!(group_commit.welcome().is_some());
+
+    // Ada joins the group and decrypts the long term secret
+    let mut ada_group = SecretGroup::new_from_welcome(&ada_provider, &group_commit).unwrap();
+    assert!(ada_group.is_active());
+
+    // ~~~~~~~~~~~~~~~~
+    // En- & Decryption
+    // ~~~~~~~~~~~~~~~~
+
+    // Billie sends an message which was encrypted with the first `LongTermSecret` to Ada, the
+    // group will automatically use the latest secret for encryption
+    let message_ciphertext = billie_group
+        .encrypt_with_long_term_secret(&billie_provider, b"This is a secret message")
+        .unwrap();
+
+    // Ada decrypts the message with the known secret
+    let message_plaintext = ada_group
+        .decrypt(&ada_provider, &message_ciphertext)
+        .unwrap();
+    assert_eq!(b"This is a secret message".to_vec(), message_plaintext);
+
+    // ~~~~~~~~~~~~
+    // Late members
+    // ~~~~~~~~~~~~
+
+    // Calvin generates a new `KeyPair` and uses it to create a `SecretGroupMember`
+    let calvin_key_pair = KeyPair::new();
+    let calvin_provider = MlsProvider::new();
+    let calvin_member = SecretGroupMember::new(&calvin_provider, &calvin_key_pair).unwrap();
+
+    // Calvin publishes their `KeyPackage` for future group invitations
+    let calvin_key_package = calvin_member.key_package(&calvin_provider).unwrap();
+
+    // Billie invites Calvin into the group
+    let group_commit = billie_group
+        .add_members(&billie_provider, &[calvin_key_package])
+        .unwrap();
+
+    // Calvin joins the group and decrypts the `LongTermSecret`
+    let mut calvin_group = SecretGroup::new_from_welcome(&calvin_provider, &group_commit).unwrap();
+    assert!(calvin_group.is_active());
+
+    // Ada processes the commit as well to sync up with the others
+    ada_group
+        .process_commit(&ada_provider, &group_commit)
+        .unwrap();
+
+    // Calvin can still decrypt the old message of Billie even though they joined the group later
+    let message_plaintext = calvin_group
+        .decrypt(&calvin_provider, &message_ciphertext)
+        .unwrap();
+    assert_eq!(b"This is a secret message".to_vec(), message_plaintext);
+
+    // Ada, Billie and Calvin still share only one long term secret
+    assert_eq!(ada_group.long_term_epoch(), Some(LongTermSecretEpoch(0)));
+    assert_eq!(billie_group.long_term_epoch(), Some(LongTermSecretEpoch(0)));
+    assert_eq!(calvin_group.long_term_epoch(), Some(LongTermSecretEpoch(0)));
+
+    // ~~~~~~~~~~~~~~~
+    // Secret rotation
+    // ~~~~~~~~~~~~~~~
+
+    // Billie removes Calvin and rotates the long term secret before to make sure Calvin will not
+    // be able to decrypt future messages
+    billie_group
+        .rotate_long_term_secret(&billie_provider)
+        .unwrap();
+    let group_commit = billie_group.remove_members(&billie_provider, &[2]).unwrap();
```

Comment:
I don't know lol, actually there is no method currently provided by `openmls` (there is a discussion in our RocketChat about it). If you check the `remove_members` docstring, there is a note on that. For now this will be it :grimacing:

Reply:
I made this :D

https://github.com/p2panda/p2panda/pull/140